pkg/sql/plan2/function/builtin/testutil.go

```diff
@@ -0,0 +1,116 @@
+// limitations under the License.
+
+package builtin
+
+import (
+	"errors"
+	"fmt"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/vectorize/eq"
+	"github.com/matrixorigin/matrixone/pkg/vm/mheap"
+	"github.com/matrixorigin/matrixone/pkg/vm/mmu/guest"
+	"github.com/matrixorigin/matrixone/pkg/vm/mmu/host"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+	"golang.org/x/exp/constraints"
+)
+
+// todo(broc): when vector.New generics are ready, change this to generic
+func MakeInt8Vector(vs []int8) *vector.Vector {
+	ret := vector.New(types.Type{Oid: types.T_int8, Size: 1})
+	vector.SetCol(ret, vs)
+	return ret
+}
+
+func MakeInt64Vector(vs []int64) *vector.Vector {
+	ret := vector.New(types.Type{Oid: types.T_int64, Size: 8})
+	vector.SetCol(ret, vs)
+	return ret
+}
+
+func MakeUInt64Vector(vs []uint64) *vector.Vector {
+	ret := vector.New(types.Type{Oid: types.T_uint64, Size: 8})
+	vector.SetCol(ret, vs)
+	return ret
+}
+
+func MakeFloat64Vector(vs []float64) *vector.Vector {
+	ret := vector.New(types.Type{Oid: types.T_float64, Size: 8})
+	vector.SetCol(ret, vs)
+	return ret
+}
+
+type TestCase struct {
+	Function        func([]*vector.Vector, *process.Process) (*vector.Vector, error)
+	InputVectors    []*vector.Vector
+	ExpectedVectors []*vector.Vector
+	ExpectError     bool
+}
+
+func RunTestCaseFloat64(testCases []TestCase) error {
+	mmu := host.New(1024 * 1024)
+	gm := guest.New(1024*1024, mmu)
+	mp := mheap.New(gm)
+	proc := process.New(mp)
+	for i, testCase := range testCases {
+		resultVector, err := testCase.Function(testCase.InputVectors, proc)
+		if err != nil {
+			if testCase.ExpectError {
+				break
+			} else {
+				return errors.New(fmt.Sprintf("unexpected error for vector %d", i))
+			}
+		}
+
+		resultValues := resultVector.Col.([]float64)
+		expectedVector := testCase.ExpectedVectors[i]
+		expectedValues := expectedVector.Col.([]float64)
+		fmt.Println(expectedValues, resultValues)
+		for j, resultValue := range resultValues {
+			if !eq.Float64Equal(resultValue, expectedValues[j]) {
+				return errors.New(fmt.Sprintf("wrong result for vector %d", i))
+			}
+		}
+	}
+	return nil
+}
+
+func RunTestCaseInteger[outputT constraints.Integer](testCases []TestCase) error {
+	mmu := host.New(1024 * 1024)
+	gm := guest.New(1024*1024, mmu)
+	mp := mheap.New(gm)
+	proc := process.New(mp)
+	for i, testCase := range testCases {
+		resultVector, err := testCase.Function(testCase.InputVectors, proc)
+		if err != nil {
+			if testCase.ExpectError {
+				break
+			} else {
+				return errors.New(fmt.Sprintf("unexpected error for vector %d", i))
+			}
+		}
+
+		resultValues := resultVector.Col.([]outputT)
+		fmt.Println(resultValues)
+		expectedVector := testCase.ExpectedVectors[i]
+		expectedValues := expectedVector.Col.([]outputT)
+		for j, resultValue := range resultValues {
```

Comment:
should consider the compare logic for Nsp of v1 and v2
eg :
b := nulls.Contain(v1.Nsp, i)
requireEqual(b, nulls.Contain(v2.Nsp, i))
if !b {
 requireEqual(v1.Cols[i], v2.Cols[i])
}